File: src/utils/chart_model_complexity.js
const QuickChart = require("quickchart-js");
const { TEST_RUNS } = require("../config");
const path = require("path");

const scenarioColors = {
  "Short Model (Low Types, Low Depth)": {
    borderColor: "rgba(74, 222, 128, 1)",
    backgroundColor: "rgba(74, 222, 128, 0.1)",
  },
  "Wide Model (Many Types, Low Depth)": {
    borderColor: "rgba(100, 180, 255, 1)",
    backgroundColor: "rgba(100, 180, 255, 0.1)",
  },
  "Deep Model (Few Types, High Depth 15)": {
    borderColor: "rgba(255, 100, 100, 1)",
    backgroundColor: "rgba(255, 100, 100, 0.1)",
  },
  "Wide & Deep Model (Many Types + High Depth)": {
    borderColor: "rgba(255, 220, 100, 1)",
    backgroundColor: "rgba(255, 220, 100, 0.1)",
  },
};

const generateCharts = async (allResults) => {
  const runLabels = Array.from({ length: TEST_RUNS }, (_, i) => `Run ${i + 1}`);
  const scenarioNames = Object.keys(allResults);

  for (const scenarioName of scenarioNames) {
    const data = allResults[scenarioName];
    const filename = `chart_model_complexity_${data.id}.png`;
    const outputPath = path.join(__dirname, "../../output", filename);

    const chart = new QuickChart();
    chart.setWidth(800);
    chart.setHeight(500);
    chart.setBackgroundColor("#1e1e2e");

    chart.setConfig({
      type: "line",
      data: {
        labels: runLabels,
        datasets: [
          {
            label: "Implicit Model ID",
            data: data.implicit,
            borderColor: "rgba(255, 100, 100, 1)",
            backgroundColor: "rgba(255, 100, 100, 0.1)",
            fill: true,
            tension: 0.3,
            pointRadius: 6,
            pointHoverRadius: 8,
          },
          {
            label: "Explicit Model ID",
            data: data.explicit,
            borderColor: "rgba(100, 180, 255, 1)",
            backgroundColor: "rgba(100, 180, 255, 0.1)",
            fill: true,
            tension: 0.3,
            pointRadius: 6,
            pointHoverRadius: 8,
          },
        ],
      },
      options: {
        title: {
          display: true,
          text: `${scenarioName} (Menor é melhor)`,
          fontColor: "#ffffff",
          fontSize: 20,
          fontFamily: "Overused Grotesk",
        },
        legend: {
          display: true,
          labels: { fontColor: "#cdd6f4", fontFamily: "Overused Grotesk" },
        },
        scales: {
          yAxes: [
            {
              ticks: {
                beginAtZero: true,
                fontColor: "#a6adc8",
                fontFamily: "Overused Grotesk",
              },
              gridLines: { color: "#333344" },
              scaleLabel: {
                display: true,
                labelString: "Latency (ms)",
                fontColor: "#6c7086",
                fontFamily: "Overused Grotesk",
              },
            },
          ],
          xAxes: [
            {
              ticks: { fontColor: "#a6adc8", fontFamily: "Overused Grotesk" },
              gridLines: { display: false },
            },
          ],
        },
      },
    });
    try {
      await chart.toFile(outputPath);
      console.log(`Generated: ${outputPath}`);
    } catch (err) {
      console.error(`Failed to generate chart for ${scenarioName}:`, err);
    }
  }
};

module.exports = { generateCharts };

File: go/pkg/utils/chart_model_complexity.go
package utils

import (
	"fmt"
)

type ModelComplexityResult struct {
	Explicit []float64
	Implicit []float64
	ID       string
	Name     string
}

var modelComplexityColors = map[string]map[string]string{
	"Modelo Curto (Poucos Tipos, Baixa Profundidade)": {
		"borderColor":     "rgba(74, 222, 128, 1)",
		"backgroundColor": "rgba(74, 222, 128, 0.1)",
	},
	"Modelo Amplo (Muitos Tipos, Baixa Profundidade)": {
		"borderColor":     "rgba(100, 180, 255, 1)",
		"backgroundColor": "rgba(100, 180, 255, 0.1)",
	},
	"Modelo Profundo (Poucos Tipos, Alta Profundidade 15)": {
		"borderColor":     "rgba(255, 100, 100, 1)",
		"backgroundColor": "rgba(255, 100, 100, 0.1)",
	},
	"Modelo Amplo e Profundo (Muitos Tipos + Alta Profundidade)": {
		"borderColor":     "rgba(255, 220, 100, 1)",
		"backgroundColor": "rgba(255, 220, 100, 0.1)",
	},
}

func GenerateModelComplexityCharts(results map[string]ModelComplexityResult, testRuns int) error {
	runLabels := make([]string, testRuns)
	for i := 0; i < testRuns; i++ {
		runLabels[i] = fmt.Sprintf("Run %d", i+1)
	}

	for scenarioName, data := range results {
		filename := fmt.Sprintf("output/chart_model_complexity_%s.png", data.ID)

		datasets := []ChartDataset{
			{
				Label:            "Implicit Model ID",
				Data:             data.Implicit,
				BorderColor:      "rgba(255, 100, 100, 1)",
				BackgroundColor:  "rgba(255, 100, 100, 0.1)",
				Fill:             true,
				Tension:          0.3,
				PointRadius:      6,
				PointHoverRadius: 8,
			},
			{
				Label:            "Explicit Model ID",
				Data:             data.Explicit,
				BorderColor:      "rgba(100, 180, 255, 1)",
				BackgroundColor:  "rgba(100, 180, 255, 0.1)",
				Fill:             true,
				Tension:          0.3,
				PointRadius:      6,
				PointHoverRadius: 8,
			},
		}

		config := ChartConfig{
			Type: "line",
		}
		config.Data.Labels = runLabels
		config.Data.Datasets = datasets
		config.Options = map[string]interface{}{
			"title": map[string]interface{}{
				"display":    true,
				"text":       fmt.Sprintf("%s (Menor é melhor)", scenarioName),
				"fontColor":  "#ffffff",
				"fontSize":   20,
				"fontFamily": "Overused Grotesk",
			},
			"legend": map[string]interface{}{
				"display": true,
				"labels": map[string]interface{}{
					"fontColor":  "#cdd6f4",
					"fontFamily": "Overused Grotesk",
				},
			},
			"scales": map[string]interface{}{
				"yAxes": []map[string]interface{}{
					{
						"ticks": map[string]interface{}{
							"beginAtZero": true,
							"fontColor":   "#a6adc8",
							"fontFamily":  "Overused Grotesk",
						},
						"gridLines": map[string]interface{}{
							"color": "#333344",
						},
						"scaleLabel": map[string]interface{}{
							"display":     true,
							"labelString": "Latency (ms)",
							"fontColor":   "#6c7086",
							"fontFamily":  "Overused Grotesk",
						},
					},
				},
				"xAxes": []map[string]interface{}{
					{
						"ticks": map[string]interface{}{
							"fontColor":  "#a6adc8",
							"fontFamily": "Overused Grotesk",
						},
						"gridLines": map[string]interface{}{
							"display": false,
						},
					},
				},
			},
		}

		if err := generateChart(config, filename); err != nil {
			return fmt.Errorf("failed to generate chart for %s: %w", scenarioName, err)
		}
	}

	return nil
}
